crypto/x509/v3_purp.c

```diff
@@ -471,9 +473,20 @@ static void x509v3_cache_extensions(X509 *x)
     if (!X509_NAME_cmp(X509_get_subject_name(x), X509_get_issuer_name(x))) {
         x->ex_flags |= EXFLAG_SI;
         /* If SKID matches AKID also indicate self signed */
-        if (X509_check_akid(x, x->akid) == X509_V_OK &&
-            !ku_reject(x, KU_KEY_CERT_SIGN))
-            x->ex_flags |= EXFLAG_SS;
+        if (X509_check_akid(x, x->akid) == X509_V_OK) {
+            /* check if the signature alg matches the PUBKEY alg */
+            X509_PUBKEY *xpkey = X509_get_X509_PUBKEY(x);
+            const X509_ALGOR *sigalg = X509_get0_tbs_sigalg(x);
+            if (xpkey != NULL && sigalg != NULL) {
+               EVP_PKEY *pkey = X509_PUBKEY_get0(xpkey);
+               int pkey_nid;
+               if (pkey != NULL
+                       && OBJ_find_sigid_algs(OBJ_obj2nid(sigalg->algorithm),
+                                              NULL, &pkey_nid)
+                       && EVP_PKEY_type(pkey_nid) == EVP_PKEY_base_id(pkey))
```

Comment:
Just as a side note, we do usually not check the signature of a self-signed cert,
unless we have X509_V_FLAG_CHECK_SS_SIGNATURE.
And since checking a signature is not for free, we should avoid checking it twice.

Reply:
There's a difference between `checking` as-in failing when the check fails, and checking whether a certificate is self-signed, just to note that fact. And the performance impact is generally low, but likely affect loading CAfiles, since we'd be checking the self-signatures on each root CA.

Reply:
Note the execution time for 1x RSA signature checking is 1/10th of a signature generation,
but the execution time for 1x EC signature checking is almost twice the execution time for
a signature generation or ECDH op.

Reply:
> If we're really out to correctly detect self-signed certs, should we perhaps check the signature here? That is just call `X509_verify()`?

At this point I'd say the given practice not to actually verify the signature is justified because this would be an expensive operation, `x509v3_cache_extensions()` appears to be called on each and any cert, also in situations where no actual cert (and chain) verification is performed, and as (correctly) stated in at least two documentation files, e.g. in `doc/man1/openssl-verify.pod.in`:
```
=item B<-check_ss_sig>

Verify the signature on the self-signed root CA. This is disabled by default
because it doesn't add any security.

```
At the point in `x509_vfy.c` where a cert (chain) is verified and the self-sig check is performed when the flag  `X509_V_FLAG_CHECK_SS_SIGNATURE` is set (which is not the default) there is a similar comment:
```
         * Skip signature check for self signed certificates unless explicitly
         * asked for because it does not add any security and just wastes time.
```

Reply:
After the strengthening of the checks in `x509v3_cache_extensions()` as I propose in this PR I'd say that the flag `EXFLAG_SS` is cached after everything that could exclude the existence of a self-signature has been checked negatively, so this is a pretty strong indication that the self-signature is there. And this indication is strong enough to work for all test cases, including the ones corrected/strengthened/added in this PR to `25-test_verify.t` and `25-test_x509.t`.
So I also do not see a pragmatic need to actually verify alleged self-signatures at this point.